### DISCUSSION TITLE
UEFI support/uefi-rs crate implementation

### DISCUSSION THREAD

**Discussion by nintendaii** [phil-opp/blog_os#1172] [Questions] [answered]
Hey. I am planning to run the OS on my UEFI laptop so I was wondering how can I add support for UEFI. As I understood, in the released version of the blog_os the whole kernel work only on BIOS. Although, I saw the uefi-rs crate on the monthly update and the name of uefi-rs crate appears from time to time in the blog_os "ecosystem", I didn’t found any specific posts about this crate implementation. Can somebody explain to me how can I implement this crate to my fork of the blog_os?

P.S. Thanks for your blog, I really appreciate what u guys are doing

**Comment by bjorn3** [comment, marked as answer]
A third edition is in the works which uses a new version of bootloader that supports both bios and uefi. Note that the bootloader crate calls ExitBootServices before passing control to the OS, so you won't be able to interact with UEFI from the OS, but it does allow booting using UEFI.

**Reply by nintendaii** [reply]
1. Does that mean that in order to simply boot the OS from my laptop with UEFI, I have to implement the uefi-rs crate using [these instructions](https://github.com/rust-osdev/uefi-rs/blob/main/BUILDING.md)? 
2. Should I wait until you guys come up with the third edition in order to properly boot from UEFI? 
3. Are here some other ways how can I boot the OS from UEFI devices?

**Reply by bjorn3** [reply]
> Does that mean that in order to simply boot the OS from my laptop with UEFI, I have to implement the uefi-rs crate using [these instructions](https://github.com/rust-osdev/uefi-rs/blob/main/BUILDING.md)?

You can find the documentation at https://docs.rs/bootloader/latest/bootloader/

> Are here some other ways how can I boot the OS from UEFI devices?


Most UEFI systems come with support for booting BIOS OSes. If not you could use something like SeaBIOS which emulate a BIOS. I think https://www.seabios.org/downloads/bios.bin.elf-1.16.0.gz is the right file for starting SeaBIOS using UEFI.

**Reply by nintendaii** [reply]
Got it. Is that correct that in order to boot the OS from BIOS/UEFI I should place the kernel image located at `target/x86-64-<os_name>/debug/bootimage-<os_name>.bin` file into USB?

**Reply by bjorn3** [reply]
I believe you have to write it as raw image to the USB using for example dd (unix) or etcher (windows) not just place it as a file.

**Reply by nintendaii** [reply]
Okay, thanks
